KAFKA-8199: Implement ValueGetter for Suppress

vvcephei opened the pull request:
See also https://github.com/apache/kafka/pull/6684

KTable processors must be supplied with a KTableProcessorSupplier, which in turn requires implementing a ValueGetter, for use with joins and groupings.

For suppression, a correct view only includes the previously emitted values (not the currently buffered ones), so this change also involves pushing the Change value type into the suppression buffer's interface, so that it can get the prior value upon first buffering (which is also the previously emitted value).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
implement ValueGetter for Suppress

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -391,7 +391,7 @@ public String queryableStoreName() {
         final String name;
         if (suppressed instanceof NamedSuppressed) {
-            final String givenName = ((NamedSuppressed) suppressed).name();
+            final String givenName = ((NamedSuppressed<?>) suppressed).name();
```
just resolving a rawtypes warning

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -402,8 +402,11 @@ public String queryableStoreName() {
+        final ProcessorSupplier<K, Change<V>> suppressionSupplier = new KTableSuppressProcessorSupplier<>(
+            suppressedInternal,
+            storeName,
+            this
```
adding a reference to the source KTable

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:
```diff
@@ -624,7 +627,7 @@ public String queryableStoreName() {
 
     @SuppressWarnings("unchecked")
-    KTableValueGetterSupplier<K, V> valueGetterSupplier() {
+    public KTableValueGetterSupplier<K, V> valueGetterSupplier() {
```
the suppress processor is in a different package.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableSourceValueGetterSupplier.java:
```diff
@@ -37,7 +37,7 @@
 
     private class KTableSourceValueGetter implements KTableValueGetter<K, V> {
-        TimestampedKeyValueStore<K, V> store = null;
+        private TimestampedKeyValueStore<K, V> store = null;
```
no need to expose this field

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java:
```diff
@@ -1,133 +0,0 @@
-
-import static java.util.Objects.requireNonNull;
-
-public class KTableSuppressProcessor<K, V> implements Processor<K, Change<V>> {
```
Moved this class inside of `KTableSuppressProcessorSupplier`

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorSupplier.java:
```diff
@@ -0,0 +1,212 @@
+    }
+
+    @Override
+    public KTableValueGetterSupplier<K, V> view() {
```
Here's the new value getter implementation.

For any data that's currently buffered, the upstream state store is actually _not_ what we want to return. For a key, we have to return the upstream state from immediately before the first record we buffered. This is now provided by the buffer, as `providedValueIfBuffered`.

For any data that's _not_ currently buffered, then we can just delegate to the parent store, since the parent table's state is equal to the suppressed table's state for this key.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorSupplier.java:
```diff
@@ -0,0 +1,212 @@
+                    @SuppressWarnings("unchecked")
+                    @Override
+                    public void init(final ProcessorContext context) {
+                        parentGetter.init(context);
```
apparently, we have to initialize the parent getter?

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorSupplier.java:
```diff
@@ -0,0 +1,212 @@
+                    @Override
+                    public void init(final ProcessorContext context) {
+                        parentGetter.init(context);
+                        // the main processor is responsible for the buffer's lifecycle
```
making a note that we should neither initialize nor close the buffer from within the getter.

vvcephei committed:
fix prior value fetch

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorSupplier.java:
```diff
@@ -0,0 +1,211 @@
+                final String[] stores = new String[1 + parentStores.length];
+                System.arraycopy(parentStores, 0, stores, 1, parentStores.length);
+                stores[0] = storeName;
+                return stores;
```
Any processor that needs to use this valuegetter needs to be connected to both the buffer and the parent table's store.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorSupplier.java:
```diff
@@ -0,0 +1,211 @@
+        parentKTable.enableSendingOldValues();
+    }
+
+    private static final class KTableSuppressProcessor<K, V> implements Processor<K, Change<V>> {
```
This class is almost verbatim from before. There are a few changes related to pushing the value type (Change<V>) into the buffer.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorSupplier.java:
```diff
@@ -0,0 +1,211 @@
+            suppressionEmitSensor = Sensors.suppressionEmitSensor(internalProcessorContext);
+
+            buffer = requireNonNull((TimeOrderedKeyValueBuffer<K, V>) context.getStateStore(storeName));
+            buffer.setSerdesIfNull((Serde<K>) context.keySerde(), (Serde<V>) context.valueSerde());
```
This, for example, used to be a fullChangeSerde, but now we just need to set the value serde.

vvcephei committed:
fix serde passed to buffer builder

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/BufferKey.java:
```diff
@@ -0,0 +1,72 @@
+
+import java.util.Objects;
+
+public final class BufferKey implements Comparable<BufferKey> {
```
Moved this out. Once I added BufferValue also, it got a little crowded in the buffer class.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.Arrays;
+import java.util.Objects;
+
+public final class BufferValue {
```
new class, since we need to _also_ carry along the prior state.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -54,17 +57,19 @@
-        new RecordHeaders(new Header[] {new RecordHeader("v", new byte[] {(byte) 1})});
+        new RecordHeaders(new Header[]{new RecordHeader("v", new byte[]{(byte) 1})});
+    private static final RecordHeaders V_2_CHANGELOG_HEADERS =
+        new RecordHeaders(new Header[]{new RecordHeader("v", new byte[]{(byte) 2})});
```
Great. We already need a new version. :((

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -77,7 +82,7 @@
     private int partition;
 
-    public static class Builder<K, V> implements StoreBuilder<StateStore> {
+    public static class Builder<K, V> implements StoreBuilder<InMemoryTimeOrderedKeyValueBuffer<K, V>> {
```
just cleanup

guozhangwang replied:
Nice cleanup!

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -146,57 +151,14 @@ public String name() {
         }
     }
 
-    private static final class BufferKey implements Comparable<BufferKey> {
```
just moved out.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -339,15 +301,18 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
+                                    record.headers()
+                                )
+                            ),
+                            null
```
Not sure the best default "prior" value... `null` removes it from the view, so that might not be great...
Could also default to the "old" value, but we'd have to deserialize it...

ableegoldman replied:
Much as it sucks to deserialize it does seem like the "correct" thing to do here (and below). That said maybe I'm misunderstanding you because I don't see why you'd have to deserialize it. Can you elaborate?

vvcephei replied:
After reflection, I agree with you. I'll make the change, and then we can discuss further.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -356,7 +321,15 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
                         time,
                         key,
-                        contextualRecord
+                        new BufferValue(contextualRecord, null)
```
same default problem here.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -424,23 +399,60 @@ public void evictWhile(final Supplier<Boolean> predicate,
     }
 
+    @Override
+    public boolean hasKey(final K key) {
```
The upstream will send us a "null" oldValue with the first update to any key. In that case, the correct view result is actually `null`, so we can't just assume `null` means we don't have the thing buffered, and we should look upstream.

Thus, I need both an explicit "hasKey" as well as a "get the prior value for this key (which might be null)"  method.

vvcephei committed:
safer prior value impl

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressScenarioTest.java:
```diff
@@ -517,6 +517,180 @@ public void shouldSupportFinalResultsForSessionWindows() {
     }
 
+    @Test
+    public void shouldWorkBeforeGroupBy() {
```
Added tests verifying the thing actually works in the operators that depend on it.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedKeyValueBufferTest.java:
```diff
@@ -147,7 +150,8 @@ public void shouldRemoveData() {
         putRecord(buffer, context, 0L, 0L, "asdf", "qwer");
         assertThat(buffer.numRecords(), is(1));
-        buffer.evictWhile(() -> true, kv -> { });
+        buffer.evictWhile(() -> true, kv -> {
```
hmm. Need to fix this.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/state/internals/TimeOrderedKeyValueBufferTest.java:
```diff
@@ -335,59 +342,64 @@ public void shouldRestoreOldFormat() {
+
+        final byte[] doomedValue = serializer.serialize(null, new Change<>("doomed", null));
+        final byte[] qwerValue = serializer.serialize(null, new Change<>("qwer", null));
+        final byte[] eo4imValue = serializer.serialize(null, new Change<>("eo4im", null));
```
This is ok, since we have a prior knowledge that the actual value type has always been Change.

vvcephei replied:
I do need to add a test making sure we can convert from the v1 format, though.

vvcephei replied:
this is done now

vvcephei reviewed (commented):
@guozhangwang @ouertani @mjsax @bbejeck @ableegoldman ,

Do (some of) you mind taking a look at this fix? I believe we should try to get it in for 2.3, and then possibly backport as well.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorSupplier.java:
```diff
@@ -0,0 +1,211 @@
+
+                    @Override
+                    public void close() {
+                        // the main processor is responsible for the buffer's lifecycle
```
Should you close the parentGetter here?

vvcephei replied:
good catch!

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -424,23 +400,60 @@ public void evictWhile(final Supplier<Boolean> predicate,
+            // it's unfortunately not possible to know this, unless we materialize the suppressed result, since our only
+            // knowledge of the prior value is what the upstream processor sends us as the "old value" when we first
+            // buffer something.
+            return ValueAndTimestamp.make(deserialize, RecordQueue.UNKNOWN);
```
Hm. How important is it to get this timestamp right? It seems like it might be necessary to store the old timestamp in addition to the old value...

vvcephei replied:
Offhand, I'd say not that important.

Many of the other valuegetters in the KTable API also default to "unknown" == -1 at the moment.

It means we have weaker-than-desired time semantics, but since we have no defined semantics at this point, it's not actually violating anything.

The solution to this is quite large in scope, so I'm leaning toward keeping this as-is for now and fixing it in a later version.

ableegoldman commented:
@vvcephei This looks good in general, I agree with your approach here modulo a few comments (+ a number of related test failures, I assume you're still getting to these?) I do feel strongly that the old value should be used when restoring from older versions, and somewhat less strongly think it might be necessary to also store the old timestamp as well. WDYT?

vvcephei commented:
Thanks @ableegoldman , I'll take a look at your comments.

* RE failures: Correct, I didn't run the tests before sending this PR, so those test failures aren't shocking. I'll take a look.
* RE restoring old values from older versions: Yeah, I think this could work. Upon reflection, `null` seems to be very much the wrong thing to use. The only redeeming factor is that buffer entries are by nature ephemeral, so any awkward default we use won't be a problem for long. Still, it seems like the old-value is a better choice.
* Re: the timestamp, If you're talking about storing the timestamp of the "prior" value, we sadly don't have access to it. I have an idea to add it, but it would require refactoring the way that records are forwarded by all KTable processors. I'd like to do this, but I don't know if it's appropriate to just slap it in right now. Should I just make a ticket to do this later?

guozhangwang commented:
Unit test failures seem relevant?

vvcephei committed:
cr comments and fix tests

vvcephei commented:
Thanks for the reviews. I've addressed the comments and fixed the tests.

I'm going to write some more tests while waiting for the next round of reviews.

vvcephei committed:
unused import

vvcephei committed:
cover parent getter in tests

vvcephei committed:
more tests

vvcephei committed:
license

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -375,42 +341,61 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
+        if (index.containsKey(key)) {
+            priorValue = internalPriorValueForBuffered(key);
+        } else {
+            final Change<V> change = valueSerde.deserializer().deserialize(
```
I don't follow this. In the case we're just loading it into the buffer from the changelog for the first time, we actually do just want oldValue to be null right? We shouldn't need to deserialize at all, just check index.containsKey

vvcephei replied:
no, when we're observing a key from the changelog that's not already buffered, it's the same as if the processor is buffering a key for the first time. The `oldValue` is set to whatever the previous state of the KTable was before the change we're suppressing. As of the instant before we suppress an update, the parent KTable and the suppressed KTable have the same view of the record. This is why (in both restoration and in processing) we can interpret the oldValue of the first change as the "prior state".

Does that make sense? It sort of still feels confusing, but I'm for some reason finding it hard to express this idea clearly :/

ableegoldman replied:
After discussing this I'm in agreement here. It's helpful to note that the serializedValue is in fact a serialized Change -- not sure if renaming this would lessen or add to the confusion...

vvcephei replied:
Hmm, tentatively, I think that renaming wouldn't be possible. The field in question is the "value" field of ContextualRecord, which is only always a change in here because that's what we put in there. Elsewhere in Streams, it can be anything in the "value" field.

ableegoldman replied:
I meant just the local variables here, not the field itself

vvcephei replied:
Ah. Sorry. I've made the change.

vvcephei committed:
verify prior value inference

vvcephei committed:
formatting

vvcephei committed:
cr feedback

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/suppress/SuppressSuite.java:
```diff
@@ -0,0 +1,42 @@
+import org.junit.runners.Suite;
+
+@RunWith(Suite.class)
+@Suite.SuiteClasses({
```
Is this for your local debugging only for added intentionally?

vvcephei replied:
I added it so that I can run all the suppression tests together. I thought it might be an interesting pattern to adopt for other functional areas. WDYT? I'm happy to remove it if you don't think it has value.

guozhangwang replied:
I actually like them and would love to keep them here :) but maybe it's better to also update the javadoc (more generally speaking, in the future if this becomes a common pattern update README as well) how to run it from gradle cmd?

vvcephei commented:
Retest this, please.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/BufferValue.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    byte[] serialize() {
+        final byte[] serializedRecord = record.serialize();
```
I'd suggest we trade a bit of code succinctness for less byte array allocation here, since this code is sort of on the critical path: we've seen that byte allocation can contribute to a large amount of CPU GC cost (e.g. https://github.com/apache/kafka/pull/6785). Here, we can first read out the length of the record plus the overhead, and then along with other two length allocate the byte array once, and then do serialize once.

vvcephei replied:
It's a little more complicated then that, since we don't know the size of the serialized RecordContext until we actually serialize it (everything else is already serialized). But I'll see what I can do...

vvcephei replied:
Addressed this by introducing a container class for the result that achieves the same effect of being able to distinguish between "absent" and `null`.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -270,22 +233,22 @@ public void flush() {
 
         final int sizeOfBufferTime = Long.BYTES;
-        final int sizeOfContextualRecord = serializedContextualRecord.length;
+        final int sizeOfBufferValue = serializedBufferValue.length;
```
I know this is not introduced in this PR, but could we consider not allocating the buffer multiple times (here we would allocate three times, each in different levels of the object's `serialize()` function).

vvcephei replied:
Nice challenge! I managed to pull this off by extending the serialization logic.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -375,42 +341,61 @@ private void restoreBatch(final Collection<ConsumerRecord<byte[], byte[]>> batch
+        if (index.containsKey(key)) {
+            priorValue = internalPriorValueForBuffered(key);
+        } else {
+            final Change<V> change = valueSerde.deserializer().deserialize(
```
Similarly here, we deserialized the whole `Changed` object but we only need the old-value here, since I know exactly how `ChangedSerde` works can we skip deserializing the new-value?

vvcephei replied:
Good call here as well. I added a utility method to the FullChangeSerde for this.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessorSupplier.java:
```diff
@@ -0,0 +1,212 @@
+
+                    @Override
+                    public ValueAndTimestamp<V> get(final K key) {
+                        if (buffer.hasKey(key)) {
```
We serialized the key twice in `hasKey` and in `priorValueForBuffered`, could we alter the interface to takes in the serialized key directly so we only do that once?

vvcephei replied:
Yep!

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -424,23 +409,69 @@ public void evictWhile(final Supplier<Boolean> predicate,
+        final Bytes serializedKey = Bytes.wrap(keySerde.serializer().serialize(changelogTopic, key));
+        final byte[] serializedValue = internalPriorValueForBuffered(serializedKey);
+
+        final V deserialize = valueSerde.innerSerde().deserializer().deserialize(
```
nit: `deserializedValue`.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java:
```diff
@@ -424,23 +409,69 @@ public void evictWhile(final Supplier<Boolean> predicate,
+        // it's unfortunately not possible to know this, unless we materialize the suppressed result, since our only
+        // knowledge of the prior value is what the upstream processor sends us as the "old value" when we first
+        // buffer something.
+        return ValueAndTimestamp.make(deserialize, RecordQueue.UNKNOWN);
```
Can we use the `bufferValue.ContextualRecord.recordContext.timestamp` inside `internalPriorValueForBuffered` here?

vvcephei replied:
Unfortunately, that would be the timestamp of the "new value", but what we need is the timestamp of the "old value". Similar to what we discussed on the KIP-258 implementation, it's not super clear what the best surrogate value would be, but it seems better to guess at a lower-than-correct time, instead of a higher-than-correct time, so that we can at least let the other side of the join "win" and set the join result timestamp. But this is a short-term pragmatic decision.

I've been mulling over this, and I think I'd like to make a jira to allow forwarding the "old value" timestamp in the DSL implementation processors.

guozhangwang replied:
Ack, that makes sense.

guozhangwang reviewed (commented):
Made a pass on the PR. Overall LGTM. One meta comment is trying to reduce redundant byte array allocation and serde as possible since most of them are on the critical path.

vvcephei committed:
remove redundant key serialization in buffer

vvcephei committed:
license

vvcephei committed:
unused import

vvcephei committed:
allocate just one buffer in ByteValue serialization

vvcephei committed:
extend single-allocation to the changelog format

vvcephei committed:
avoid round-trip to get serialized prior value

vvcephei committed:
nit

vvcephei committed:
fix

vvcephei commented:
@guozhangwang , thanks for the review. I addressed the spots you pointed out. Some further optimization is possible, but it would require swapping out the `byte[]` in ContextualRecord in favor of a ByteBuffer and then carefully managing the cursors. This way, we could achieve the de-duplication @ableegoldman  was mentioning, by making the "prior value" just a view into the serialized value when it happens to be the same as the "old value".

But this seems potentially more risky than it's worth... We have to be careful to always re-set the buffer's cursor when we're done reading from it (and it wouldn't be thread safe, which might be hard to remember if we need to make this collection thread-safe in the future). Also, when we get an update, the serialized value would be updated, but the "prior value" would keep alive a reference to the whole originalized serialized value, even though we only need the old-value part. We would have to have an extra step to allocate the independent "prior value" array when we first overwrite the value (and then remember not to do it again for the following updates).

Neither of these are insurmountable problems, but I wanted to run it by you before taking on the involved complexity. Do you think it's worth it, or have we done enough by eliminating the extra allocations you already pointed out?

Oh, also, do you want me to remove the test suite I added?

Thanks again!
-John

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/Maybe.java:
```diff
@@ -0,0 +1,89 @@
+ *
+ * @param <T>
+ */
+public final class Maybe<T> {
```
Hmm.. can we use `Optional` instead?

vvcephei replied:
No, since Optional cannot contain `null`, but that's specifically what we want to do here (represent both `null` and "empty" independently). We _could_ use `Optional<Optional<T>>`, but that just seems silly ;)

guozhangwang replied:
Ack.

guozhangwang reviewed (commented):
A couple more comments, otherwise LGTM>

guozhangwang commented:
retest this please

guozhangwang commented:
I think we can live with this for now, if it becomes a GC issue we can then
consider trading more code complexity then (maybe file a JIRA just in case)?

On Wed, May 29, 2019 at 5:48 PM John Roesler <notifications@github.com>
wrote:

> @guozhangwang <https://github.com/guozhangwang> , thanks for the review.
> I addressed the spots you pointed out. Some further optimization is
> possible, but it would require swapping out the byte[] in
> ContextualRecord in favor of a ByteBuffer and then carefully managing the
> cursors. This way, we could achieve the de-duplication @ableegoldman
> <https://github.com/ableegoldman> was mentioning, by making the "prior
> value" just a view into the serialized value when it happens to be the same
> as the "old value".
>
> But this seems potentially more risky than it's worth... We have to be
> careful to always re-set the buffer's cursor when we're done reading from
> it (and it wouldn't be thread safe, which might be hard to remember if we
> need to make this collection thread-safe in the future). Also, when we get
> an update, the serialized value would be updated, but the "prior value"
> would keep alive a reference to the whole originalized serialized value,
> even though we only need the old-value part. We would have to have an extra
> step to allocate the independent "prior value" array when we first
> overwrite the value (and then remember not to do it again for the following
> updates).
>
> Neither of these are insurmountable problems, but I wanted to run it by
> you before taking on the involved complexity. Do you think it's worth it,
> or have we done enough by eliminating the extra allocations you already
> pointed out?
>
> Oh, also, do you want me to remove the test suite I added?
>
> Thanks again!
> -John
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/kafka/pull/6781?email_source=notifications&email_token=ACR3Q7JFZP73FUX7M3BUPZLPX4P47A5CNFSM4HOOEZKKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWRAYOI#issuecomment-497159225>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACR3Q7N4CD3D2PQYSL5HJ33PX4P47ANCNFSM4HOOEZKA>
> .
>


-- 
-- Guozhang

vvcephei commented:
Thanks, @guozhangwang ! 

There was just one unrelated test failure in the java 11 build:
* https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/5108/ `kafka.integration.UncleanLeaderElectionTest.testUncleanLeaderElectionDisabled`

I'll add the javadoc and README you mentioned on the test suite, which will have the side-effect of re-running the tests.

I'll also create the Jira you mentioned.

vvcephei committed:
add javadoc for SuppressSuite

guozhangwang commented:
LGTM. Wait for green builds.

vvcephei commented:
While writing the Jira, I had a better idea for the optimization... I'll send a PR against this PR so you can see what you think independently. Give me a sec...

bbejeck reviewed (approved):
Sorry for the late review.  I don't have anything to add beyond what @guozhangwang and @ableegoldman have commented on.

vvcephei commented:
Thanks for the reviews, all.

I (finally) had a decent idea about how to deduplicate the prior value (https://github.com/vvcephei/kafka/pull/1). This is an optional add-on, so we can just merge this PR now, and make https://github.com/vvcephei/kafka/pull/1 a separate PR (with a separate Jira). It should be backward compatible with the v2 changelog wire format, so there's no strong reason to do them both together. Just wanted to give you the option.

guozhangwang commented:
LGTM! Thanks for the patch @vvcephei

guozhangwang commented:
Cherry-picked to 2.3 as well.
